R/utils.R

```diff
@@ -163,3 +165,108 @@ h_tr <- function(x) {
   }
   sum(diag(x))
 }
+
+#' Split Control List
+#' @description Split the [mmrm_control()] object according to its optimizers and use additional arguments
+#' to replace the elements in the original object.
+#'
+#' @param control (`mmrm_control`)\cr object.
+#' @param ... additional parameters to update the `control` object.
+#'
+#' @keywords internal
+h_split_control <- function(control, ...) {
+  assert_class(control, "mmrm_control")
+  l <- length(control$optimizers)
+  lapply(seq_len(l), function(i) {
+    ret <- modifyList(control, list(...))
+    ret$optimizers <- control$optimizers[i]
+    ret
+  })
+}
+
+#' Obtain Optimizer according to Optimizer String Value
+#' @description This function creates optimizer functions with arguments.
+#'
+#' @param optimizer (`character`)\cr optimizer character. Only "L-BFGS-B", "BFGS", "CG" and "nlminb" supported.
+#' @param optimizer_fun (`function`)\cr or (`list`)\cr of (`function`)x\cr.
+#' @param optimizer_args (`list`)\cr of additional arguments for `optimizer_fun`.
+#' @param optimizer_control (`list`)\cr of control argument. Will serve as `control` in `optimizer_fun`.
+#'
+#' @details
+#' In this function, the final object to use is `optimizer_fun`, and `optimizer` is a shortcut to create
+#' multiple `optimizer_fun`. If you want to use your own optimizer function, make sure that there are three arguments,
+#' parameter(start value), objective function and gradient function are sequentially in the function arguments.
+#' If there are other named arguments in front of these, make sure they are correctly specified through
+#' `optimizer_args`. If hessian function can be applied, please make sure its argument name is `hessian` and
+#' please add attribute `use_hessian = TRUE` to the function, using `attr(fun, "use_hessian) <- TRUE`
+#' If `optimizer_fun` is not provided, it will be created using the `optimizer` character. Other arguments should
+#' go into `optimizer_args`.
+#' Allowed are "L-BFGS-B", "BFGS", "CG"(using [stats::optim()] with corresponding method) and
+#' "nlminb"(using [stats::nlminb()]).
+#'
+#' @keywords internal
+h_get_optimizers <- function(optimizer = c("L-BFGS-B", "BFGS", "CG", "nlminb"),
```

Comment:
what happens if user supplies `automatic`? can we fail with a deprecation message instead of just somehow?